near/omni-prover/evm-prover/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+        let data = Self::verify_trie_proof(
+            header.receipts_root.0,
+            rlp::encode(&evm_proof.receipt_index).to_vec(),
+            &evm_proof.proof,
+        );
+
+        if evm_proof.receipt_data != data {
+            return Err("ERR_INVALID_PROOF".to_owned());
+        }
+
+        // Verify block header was in the bridge
+        Ok(evm_client::ext(self.light_client.clone())
+            .with_static_gas(BLOCK_HASH_SAFE_GAS)
+            .block_hash_safe(header.number.as_u64())
+            .then(
+                Self::ext(env::current_account_id())
+                    .with_static_gas(VERIFY_PROOF_CALLBACK_GAS)
+                    .verify_proof_callback(
+                        args.proof_kind,
+                        evm_proof.log_entry_data,
+                        header.hash.ok_or("ERR_HASH_NOT_SET")?.0,
+                    ),
+            ))
+    }
+
+    #[private]
+    #[handle_result]
+    pub fn verify_proof_callback(
+        &mut self,
+        #[serializer(borsh)] kind: ProofKind,
+        #[serializer(borsh)] log_entry_data: Vec<u8>,
+        #[serializer(borsh)] expected_block_hash: H256,
+        #[callback]
+        #[serializer(borsh)]
+        block_hash: Option<H256>,
+    ) -> Result<ProverResult, String> {
+        if block_hash != Some(expected_block_hash) {
+            return Err("ERR_INVALID_BLOCK_HASH".to_owned());
+        }
+
+        match kind {
+            ProofKind::InitTransfer => Ok(ProverResult::InitTransfer(parse_evm_event(
+                self.chain_kind,
+                log_entry_data,
+            )?)),
+            ProofKind::FinTransfer => Ok(ProverResult::FinTransfer(parse_evm_event(
+                self.chain_kind,
+                log_entry_data,
+            )?)),
+            ProofKind::DeployToken => Ok(ProverResult::DeployToken(parse_evm_event(
+                self.chain_kind,
+                log_entry_data,
+            )?)),
+        }
+    }
+
+    /// Verify the proof recursively traversing through the key.
+    /// Return the value at the end of the key, in case the proof is valid.
+    ///
+    /// @param expected_root is the expected root of the current node.
+    /// @param key is the key for which we are proving the value.
+    /// @param proof contains relevant information to verify data is valid
+    ///
+    /// Patricia Trie: https://eth.wiki/en/fundamentals/patricia-tree
+    /// Patricia Img:  https://ethereum.stackexchange.com/questions/268/ethereum-block-architecture/6413#6413
+    ///
+    /// Verification:  https://github.com/slockit/in3/wiki/Ethereum-Verification-and-MerkleProof#receipt-proof
+    /// Article:       https://medium.com/@ouvrard.pierre.alain/merkle-proof-verification-for-ethereum-patricia-tree-48f29658eec
+    /// Python impl:   https://gist.github.com/mfornet/0ff283274c0162f1cca45966bccf69ee
+    ///
+    fn verify_trie_proof(expected_root: H256, key: Vec<u8>, proof: &Vec<Vec<u8>>) -> Vec<u8> {
```

Comment:
We must support the old prover as well. This function is an exact copy-paste. In the short term, it’s easier to do it this way, and it also avoids unnecessary dependencies. However, it might make sense in the future to move this code to a separate library."

Reply:
Yeh, agree